### DISCUSSION TITLE
Cycle through active groups in the current screen

### DISCUSSION THREAD

**Discussion by asoymos** [qtile/qtile#4722] [Q&A] [answered]
Hi and thank you. I have two monitors with different groups for each as specified in [FAQ](https://docs.qtile.org/en/stable/manual/faq.html#how-can-i-get-my-groups-to-stick-to-screens) and would be very useful to have a keyboard shorcut to cycle through the active groups only in the current screen. I have searched, read documentation, examples and discussions, but I am new to python and haven't found something simple that I can use. Any recommendations?

**Comment by elParaguayo** [comment]
Something like this:
```python
@lazy.function
def loop_screen_groups(qtile):
    # Get the current screen index
    screen = qtile.current_screen.index
    # Find all groups marked for that screen or that have no affinity
    screen_groups = [g for g in qtile.groups if g.screen_affinity in [screen, None]]
    # Get the index of the next group
    current = screen_groups.index(qtile.current_group)
    next_index = (current + 1) % len(screen_groups)
    # Show that group
    screen_groups[next_index].toscreen()
```

**Reply by asoymos** [reply]
Thank you. It works well but it cycle through all groups of a screen, it is posible to cycle only active groups?

**Reply by elParaguayo** [reply, marked as answer]
What do you mean by active? That the group has windows? If so, change this line:
```python
screen_groups = [g for g in qtile.groups if g.screen_affinity in [screen, None] and g.windows]
```

**Reply by asoymos** [reply]
Exactly that. Only thing now is that it can't be used to switch from an empty group, log complain: `ValueError: <group.Group ('3')> is not in list`. But it's very much what I needed, I will try to learn more and improve it. Thank you!

**Reply by elParaguayo** [reply]
My code doesn't include any error handling. It would be easy enough to add some to address that situation.

**Reply by asoymos** [reply]
(sorry for any lenguage mistakes, english is not my first lenguage) I am trying to make a conditional to detect `if qtile.current_group in screen_groups.index:` but log output `argument of type 'builtin_function_or_method' is not iterable` so how I can know if the current group is in screen_groups?

What I have is this:
```python
@lazy.function
def loop_screen_active_groups(qtile):
    # Get the current screen index
    screen = qtile.current_screen.index
    # Find all groups marked for that screen or that have no affinity and that have windows
    screen_groups = [g for g in qtile.groups if g.screen_affinity in [screen, None] and g.windows]
    # si el grupo actual está en screen_groups
    if qtile.current_group in screen_groups.index:
        # Get the index of the next group
        current = screen_groups.index(qtile.current_group)
        next_index = (current + 1) % len(screen_groups)
        # Show that group
        screen_groups[next_index].toscreen()
    else:
        for group in screen_groups:
            qtile.current_screen.set_group(group)
            break
```

**Reply by asoymos** [reply]
(haven't read your last comment) Or how do you recommend to address the situation that if I am in an empty (no windows) group go to next active (with windows) group?

**Reply by elParaguayo** [reply]
First off, your English is great - please don't apologise!

> I am trying to make a conditional to detect if qtile.current_group in screen_groups.index: but log output argument of type 'builtin_function_or_method' is not iterable so how I can know if the current group is in screen_groups?

`.index` returns a number specifying the index of an object in a list. If you want to check if a group is in the list, just do `if qtile.current_group in screen_groups:`

> Or how do you recommend to address the situation that if I am in an empty (no windows) group go to next active (with windows) group?

Your if statement does that already.

**Reply by asoymos** [reply]
Yeah it's working now! Gracias!!

**Reply by elParaguayo** [reply]
De nada!

**Reply by asoymos** [reply]
Hi again. When I move the only window of a group and use this function to cycle through groups with windows it go to the first group of the index (eg. if I am in empty group 2 and there are group 1 and 3 with windows it to go to group 1). How can I make it to go instead to next group (group 3) and then keep cycling?

This is the actual function:
```python
@lazy.function
def loop_screen_active_groups(qtile):
    # Get the current screen index
    screen = qtile.current_screen.index
    # Find all groups marked for that screen or that have no affinity and that have windows
    screen_groups = [g for g in qtile.groups if g.screen_affinity in [screen, None] and g.windows]
    # si el grupo actual está en screen_groups
    if qtile.current_group in screen_groups:
        # Get the index of the next group
        current = screen_groups.index(qtile.current_group)
        next_index = (current + 1) % len(screen_groups)
        # Show that group
        screen_groups[next_index].toscreen()
    else:
        for group in screen_groups:
            qtile.current_screen.set_group(group)
            break
```
Thanks.

**Reply by asoymos** [reply]
I tought of copying `current = qtile.groups.index(qtile.current_group)` but with qtile.groups instead of screen_groups to get current group index, but `next_index = (current + 1) % len(screen_groups)` don't work as current needs to be inside screen_groups I think, so I don't know the math now to get to next group with windows.

**Comment by Cyber-Syntax** [comment]
I didn't quietly understand what you're trying to achieve. Anyway, I though maybe my function give you some idea to find out your problem: 
```python
# Cycle on groups.
@lazy.function
def cycle_groups(qtile):
    """Cycle through the groups but only the odd or even groups"""
    current_group_index = qtile.groups.index(qtile.current_group)
    next_group_index = current_group_index

# Check if the current group name is odd or even
    current_group_name = qtile.current_group.name
    is_odd_group = int(current_group_name) % 2 != 0

# Loop through the groups until we find the desired group
    while True:
        # Get the next group name
        next_group_index = (next_group_index + 1) % len(qtile.groups)
        next_group_name = qtile.groups[next_group_index].name

        # Check if the next group name has the desired property (odd or even)
        if is_odd_group and int(next_group_name) % 2 != 0:
            break
        
        if not is_odd_group and int(next_group_name) % 2 == 0:
            break
# Lets get back to the first group when we reach the last group
# if 5 -> 1, 6 -> 2

        if next_group_index == 6:
            # -1 = refers to the group 1, 0 refers to the group 2
            next_group_index = 0
        elif next_group_index == 5:
            next_group_index = -1


    qtile.current_screen.set_group(qtile.groups[next_group_index])
```
Also I made AI to make me reverse cycle group function from my function too:
```python
@lazy.function
def cycle_groups_reverse(qtile):
    """Cycle through the groups in reverse order but only the odd or even groups"""
    current_group = qtile.current_group
    current_group_num = int(current_group.name)
    is_odd = current_group_num % 2 != 0
    
    # Get all group numbers
    group_nums = [int(group.name) for group in qtile.groups]
    
    # Filter for odd or even groups based on the current group
    filtered_nums = [num for num in group_nums if num % 2 != 0] if is_odd else [num for num in group_nums if num % 2 == 0]
    
    # Sort in descending order
    filtered_nums.sort(reverse=True)
    
    # Find the index of the current group in our filtered list
    current_index = filtered_nums.index(current_group_num)
    
    # Get the next group number (wrapping around if necessary)
    next_group_num = filtered_nums[(current_index + 1) % len(filtered_nums)]
    
    # Find the qtile group object with this number
    next_group = next(group for group in qtile.groups if int(group.name) == next_group_num)
    
    # Switch to the next group
    next_group.toscreen()
```
